File: ruby/anagram/anagram.rb
class Anagram
  
  def initialize(word)
    @word = FormattedWord.new(word)
  end

  def match(possible_anagrams)
    possible_anagrams.select { |possible_anagram|  FormattedWord.new(possible_anagram).anagram_of?(@word) }
  end

end

class FormattedWord

  attr_reader :candidate
  def initialize(candidate)
    @candidate = candidate.downcase
  end

  def sort
    candidate.chars.sort.join
  end

  
  def anagram_of?(other_word)
    candidate != other_word.candidate && self.sort == other_word.sort
  end

end

File: elixir/anagram/anagram.exs
defmodule Anagram do
  @doc """
  Returns all candidates that are anagrams of, but not equal to, 'base'.
  """
  @spec match(String.t, [String.t]) :: [String.t]
  def match(base, candidates) do
    candidates
    |> Enum.filter(fn candidate -> is_anagram_of?(candidate, base) end)
  end

  @spec is_anagram_of?(String.t, String.t) :: boolean
  defp is_anagram_of?(candidate, base) do
   not_the_same_word?(candidate, base) && normalize_string(candidate) == normalize_string(base)
  end

  @spec not_the_same_word?(String.t, String.t) :: boolean
  defp not_the_same_word?(candidate, base) do
    String.downcase(candidate) != String.downcase(base)
  end

  @spec normalize_string(String.t) :: String.t
  defp normalize_string(string) do
    string
    |> String.downcase
    |> String.split("")
    |> Enum.sort
    |> Enum.join
  end
end
